File: java/src/sharedkernel/infrastructure/commandbus/SimpleCommandBus.java
package sharedkernel.infrastructure.commandbus;

import sharedkernel.application.command.Command;
import sharedkernel.application.command.handler.Handler;
import sharedkernel.application.commandbus.CommandBus;

import java.util.HashMap;
import java.util.Map;

public class SimpleCommandBus implements CommandBus {
    private Map<String, Handler> handlers;

    public SimpleCommandBus() {
        handlers = new HashMap<>();
    }

    @Override
    public void registerHandler(Command command, Handler handler) {
        handlers.put(command.getClass().getSimpleName(), handler);
    }

    @Override
    public void handle(Command command) {
        handlers.get(command.getClass().getSimpleName()).handle(command);
    }
}

File: kotlin/src/sharedkernel/infrastructure/commandbus/SimpleCommandBus.kt
package sharedkernel.infrastructure.commandbus

import sharedkernel.application.command.Command
import sharedkernel.application.command.handler.Handler
import sharedkernel.application.commandbus.CommandBus

import java.util.HashMap

class SimpleCommandBus : CommandBus {
    private val handlers: MutableMap<String, Handler>

    init {
        handlers = HashMap()
    }

    override fun registerHandler(command: Command, handler: Handler) {
        handlers[command.javaClass.simpleName] = handler
    }

    override fun handle(command: Command) {
        handlers[command.javaClass.simpleName]?.handle(command)
    }
}
